src/ucx_van.h

```diff
@@ -751,11 +758,25 @@ class UCXVan : public Van {
     UCX_REQUEST_FREE(req); // can release request back to UCX now
   }
 
-  ucp_tag_t MakeTag(int node_id, ucp_tag_t tag) {
-    return ((ucp_tag_t)node_id << 32) | tag;
+  /**
+   * from the left to the right:
+   * 1 bit for UCX_TAG_META/UCX_TAG_DATA
+   * 31 bits for the sender node_id
+   * 32 bits for the key
+   */
+  ucp_tag_t MakeTag(int node_id, Tags tag, uint64_t key) {
+    ucp_tag_t ret = 0;
+
```

Comment:
maybe add asserts here that sender node_id takes less than 32 bits and key takes at most 32 bits?

Reply:
I added assert to check node_id is less than 32 bits. we don't need to check for the key because we only keep the lower 32 bits no matter what.